## Demonstrating NLP methods of finding text similarities and anomaly detection

Natural Language Processing (NLP) methods for finding text similarities and performing anomaly detection are essential for various applications, including information retrieval, recommendation systems, and fraud detection.

Text Similarity with NLP:
Text similarity is the process of determining how similar two pieces of text are. Here's an example of using NLP methods for text similarity:


## Cosine Similarity:



- Cosine similarity measures the cosine of the angle between two non-zero vectors in a multi-dimensional space. It is commonly used for text similarity tasks. You can use popular NLP libraries such as spaCy or scikit-learn in Python to calculate cosine similarity.

In [9]:
# Import Libraries
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.metrics.pairwise import cosine_similarity


In [10]:
text1 = "Machine Learning is the future technology."
text2 = "Artificial Intelligence will shape the future of technology."

# Create a TF-IDF vectors
vectorizer = TfidfVectorizer()
tfidf_matrix = vectorizer.fit_transform([text1, text2])

# Display the feature names and TF-IDF matrix
print("Feature Names:", vectorizer.get_feature_names_out())
print("TF-IDF Matrix:\n", tfidf_matrix.toarray())

# Calculate cosine similarity
cosine_similarity = cosine_similarity(tfidf_matrix[0], tfidf_matrix[1])
print("Cosine Similarity: ", cosine_similarity[0][0])

Feature Names: ['artificial' 'future' 'intelligence' 'is' 'learning' 'machine' 'of'
 'shape' 'technology' 'the' 'will']
TF-IDF Matrix:
 [[0.         0.33471228 0.         0.47042643 0.47042643 0.47042643
  0.         0.         0.33471228 0.33471228 0.        ]
 [0.39166832 0.27867523 0.39166832 0.         0.         0.
  0.39166832 0.39166832 0.27867523 0.27867523 0.39166832]]
Cosine Similarity:  0.2798280652432878


TF-IDF matrix represents the weight of each word in each text. 
The cosine similarity score reflects the overall similarity between the two TF-IDF vectors.

Cosine similarity ranges from -1 to 1:

- 1: The vectors are identical. This measures the cosine of the angle between the vectors, with smaller angles (closer to 1) indicating higher similarity.
- 0: The vectors are orthogonal (no similarity).
- -1: The vectors are diametrically opposite.
- A score of 0.2798 suggests that the two texts share some common terms but are not very similar. 

TF-IDF Vector Calculation:
- Term Frequency (TF): How often a word appears in a document.
- Inverse Document Frequency (IDF): Measures the importance of a word across the corpus. Common words have lower IDF scores.
- TF-IDF: Combines TF and IDF to give a weight to each word, reflecting its importance in the document relative to the corpus.

Practical Implications
A moderate cosine similarity score can be interpreted in various ways:

- Text Comparison: Helps in identifying related documents or passages.
- Recommendation Systems: Suggests similar items based on textual descriptions.
- Information Retrieval: Enhances search engines by ranking results based on textual relevance.

## Word Embeddings (Word2Vec, GloVe):

Word embeddings are dense vector representations of words. You can calculate text similarity using pre-trained word embeddings and cosine similarity.

In [16]:
from gensim.models import Word2Vec
from sklearn.metrics.pairwise import cosine_similarity

In [21]:
!pip install gensim



[notice] A new release of pip is available: 24.0 -> 24.1
[notice] To update, run: python.exe -m pip install --upgrade pip


In [24]:
# Sample corpus
sentences = [
    ["machine", "learning", "is", "the", "future", "technology"],
    ["artificial", "intelligence", "will", "shape", "the", "future", "of", "technology"]
]

# Train Word2Vec model
model = Word2Vec(sentences, vector_size=100, window=5, min_count=1, workers=4)

# Save the model
model.save("word2vec.model")

Training the Model:

- sentences: A list of tokenized sentences (corpus) used to train the Word2Vec model.
- Word2Vec: Trains the model on the given corpus with specified parameters like vector_size, window, min_count, and workers.
- model.save("word2vec.model"): Saves the trained model to a file named word2vec.model.

In [30]:
# Load pre-trained Word2Vec model
model = Word2Vec.load("Word2Vec.model")

# Calculate cosine similarity between two sentences
similarity = cosine_similarity([model.wv['machine'], model.wv['learning']],[model.wv['artificial'], model.wv['intelligence']])

print("Cosine Similarity: ", similarity[0][0])

Cosine Similarity:  0.108892545


# Anomaly Detection with NLP


Anomaly detection in text involves identifying unusual or outlier patterns in textual data. Here's an example of using NLP for anomaly detection:

### Outlier Detection with TF-IDF:

You can use TF-IDF to detect anomalies by identifying documents that significantly differ from the others.

In [32]:
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.ensemble import IsolationForest

In [33]:
documents = ["Normal text", "Another normal text", "Anomaly detected here"]

# Create TF-IDF vectors
vectorizer = TfidfVectorizer()
tfidf_matrix = vectorizer.fit_transform(documents)

# Detect anomalies using isolation forest
clf = IsolationForest(contamination=.2)
clf.fit(tfidf_matrix.toarray())
anomalies = clf.predict(tfidf_matrix.toarray())

print("Anomalies:", anomalies)


Anomalies: [ 1  1 -1]


IsolationForest is an anomaly detection algorithm.
- contamination=0.2 specifies that approximately 20% of the data is expected to be anomalies.
- By adjusting the contamination parameter, you can control the sensitivity of the anomaly detection process, making it suitable for different use cases and datasets.
- 1 indicates normal points.
- -1 indicates anomalies.

Practical Implications
This approach can be used to automatically detect anomalous documents or texts in a corpus, which can be useful in various applications such as:

- Fraud Detection: Identifying unusual transactions or behaviors.
- Quality Control: Detecting defective products or processes.
- Security: Identifying malicious activities or intrusions.

### LSTM Autoencoder for Anomaly Detection

For more advanced anomaly detection in sequences (e.g., identifying abnormal text patterns in a sequence of sentences), you can use recurrent neural networks (RNN) like LSTM autoencoders.

In [77]:
import numpy as np
from tensorflow.keras.layers import Input, LSTM, RepeatVector, TimeDistributed, Dense
from tensorflow.keras.models import Model
from tensorflow.keras.preprocessing.text import Tokenizer
from tensorflow.keras.preprocessing.sequence import pad_sequences
from tensorflow.keras.utils import to_categorical

In [82]:

# Sample sequences
sequences = ["Normal sentence 1", "Normal sentence 2", "Anomaly sentence"]

# Tokenize and pad sequences
max_sequence_length = max(len(s.split()) for s in sequences)
print(f"max_sequence_length - {max_sequence_length}")

tokenizer = Tokenizer()
tokenizer.fit_on_texts(sequences)
sequences_tokenized = tokenizer.texts_to_sequences(sequences)
print(f"sequences - {sequences_tokenized}")

sequences_padded = pad_sequences(sequences_tokenized, maxlen=max_sequence_length, padding='post')
print(f"pad_sequences - {sequences_padded}")

# One-hot encode the padded sequences
vocab_size = len(tokenizer.word_index) + 1  # Adding 1 for the <OOV> token
sequences_one_hot = np.array([to_categorical(seq, num_classes=vocab_size) for seq in sequences_padded])
print(f"One-hot encoded sequences - {sequences_one_hot}")

# Create LSTM autoencoder
input_seq = Input(shape=(max_sequence_length, vocab_size))
encoded = LSTM(64)(input_seq)
decoded = RepeatVector(max_sequence_length)(encoded)
decoded = LSTM(64, return_sequences=True)(decoded)
output_seq = TimeDistributed(Dense(vocab_size, activation='softmax'))(decoded)

autoencoder = Model(input_seq, output_seq)
autoencoder.compile(optimizer='adam', loss='categorical_crossentropy')

# Train the autoencoder
autoencoder.fit(sequences_one_hot, sequences_one_hot, epochs=100, batch_size=1)

# Detect anomalies by measuring reconstruction error
reconstructed_sequences = autoencoder.predict(sequences_one_hot)
print('reconstructed_sequences:', reconstructed_sequences)

anomaly_scores = np.mean(np.square(sequences_one_hot - reconstructed_sequences), axis=(1, 2))
print('Anomaly scores:', anomaly_scores)

max_sequence_length - 3
sequences - [[2, 1, 3], [2, 1, 4], [5, 1]]
pad_sequences - [[2 1 3]
 [2 1 4]
 [5 1 0]]
One-hot encoded sequences - [[[0. 0. 1. 0. 0. 0.]
  [0. 1. 0. 0. 0. 0.]
  [0. 0. 0. 1. 0. 0.]]

 [[0. 0. 1. 0. 0. 0.]
  [0. 1. 0. 0. 0. 0.]
  [0. 0. 0. 0. 1. 0.]]

 [[0. 0. 0. 0. 0. 1.]
  [0. 1. 0. 0. 0. 0.]
  [1. 0. 0. 0. 0. 0.]]]
Epoch 1/100
3/3 [==============================] - 5s 6ms/step - loss: 1.7934
Epoch 2/100
3/3 [==============================] - 0s 6ms/step - loss: 1.7844
Epoch 3/100
3/3 [==============================] - 0s 11ms/step - loss: 1.7751
Epoch 4/100
3/3 [==============================] - 0s 8ms/step - loss: 1.7674
Epoch 5/100
3/3 [==============================] - 0s 7ms/step - loss: 1.7587
Epoch 6/100
3/3 [==============================] - 0s 6ms/step - loss: 1.7488
Epoch 7/100
3/3 [==============================] - 0s 7ms/step - loss: 1.7381
Epoch 8/100
3/3 [==============================] - 0s 6ms/step - loss: 1.7273
Epoch 9/100
3/3 [===============

Training Output
Epochs 1-100:

Training is carried out over 100 epochs, which means the entire training dataset is passed forward and backward through the neural network 100 times.
The loss decreases consistently over the epochs, indicating that the model is learning and improving its performance.

Loss Value:

The loss value measures how well the model's predictions match the actual data. A lower loss indicates better model performance.
The consistent reduction in loss over epochs signifies that the model is successfully learning to reconstruct the sequences.

Anomaly Scores:

- Anomaly scores: [2.728174e-06, 6.158271e-06, 1.629073e-05]
- These scores represent the reconstruction error for each sequence.
- Lower scores indicate sequences that are well-reconstructed by the autoencoder and are considered normal.
- Higher scores indicate sequences that deviate from the expected pattern and are considered anomalies.

The model shows good learning progression, with the loss decreasing significantly over 100 epochs.
Anomaly scores are very low, indicating that the model has learned to reconstruct the input sequences well, and there are no significant anomalies in the provided data.